### DISCUSSION TITLE
Call to undefined method Symfony\Component\HttpFoundation\StreamedResponse::header()

### DISCUSSION THREAD

**Discussion by jindong02** [community/community#125143] [Programming Help] [answered]
### Body

I am developing file download feature in laravel.
But I can't download files because of this error;

**Call to undefined method Symfony\Component\HttpFoundation\StreamedResponse::header()**
This is controller class.
`
namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Storage;
use Symfony\Component\HttpFoundation\StreamedResponse;
class FileDownload  extends Controller
{
    public function download(Request $request, $path)
    {
        if (!Storage::disk('uploads')->exists($path)) {
            abort(404, "The path doesn't exist.");
        }
        // Check if the path is a file
        $fullPath = storage_path("app/public/" . $path);
        if (is_dir($fullPath)) {
            abort(404, "The path is a directory, not a file.");
        }
        return Storage::disk('uploads')->response($path);
    }
}
`
If $path is wrong, 404 occurred.
But $path is right above error occurs.
Please help me I can download file correctly.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by billythekid** [comment]
Hi @jindong02 :wave:

This code looks correct to me. I'm not seeing the `StreamedResponse` class being called in your code, but it's being `use`d at the top. 

What happens if you comment out the line using the Class? 

```php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Storage;
// use Symfony\Component\HttpFoundation\StreamedResponse;

class FileDownload extends Controller
{
    public function download(Request $request, $path)
    {
        if (!Storage::disk('uploads')->exists($path)) {
            abort(404, "The path doesn't exist.");
        }
// Check if the path is a file
        $fullPath = storage_path("app/public/" . $path);
        if (is_dir($fullPath)) {
            abort(404, "The path is a directory, not a file.");
        }
        return Storage::disk('uploads')->response($path);
    }
}
```

**Reply by billythekid** [reply]
To address the error...

The [latest version of StreamedResponse](https://github.com/symfony/symfony/blob/7.2/src/Symfony/Component/HttpFoundation/StreamedResponse.php) doesn't have a header() method, nor does the Request parent class.

Whatever code is attempting to call that, is failing. Maybe it should be `sendHeaders()`?

**Reply by jindong02** [reply]
I use Laravel 11.
My project has cors middleware.
This middleware set cors option to response header.
Above Error handled here only when I request images but in other cases this middleware works well .
When I don't use StreamedResponse, also same error handle.

Here is cors middleware


`<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Symfony\Component\HttpFoundation\Response;

class Cors
{
    public function handle(Request $request, Closure $next): Response
    {
        return $next($request)
            ->header('Access-Control-Allow-Origin', '*');
    }
}`

**Comment by fdev0562** [comment, marked as answer]
StreamedResponse doesn't have header function.
If you have to use cors middleware there is one way.
You can filter some url like file download url in cors middleware so that response for download urls can't be handled in cors middleware.

Below is fixed code.

```php
public function handle(Request $request, Closure $next): Response
{
   //here 'files/*' have to be replaced by your download url.
    if ($request->is('files/*')) {
        // Skip adding the CORS headers for URLs starting with '/uploads/'
        return $next($request);
    }

    return $next($request)
        ->header('Access-Control-Allow-Origin', '*');
 }

```

**Reply by jindong02** [reply]
It works well, thank you.
I get a good idea from your answer.
